src/Tools/AnalyzerRunner/Program.cs

```diff
@@ -0,0 +1,480 @@
+                var diagnostics = await GetAnalyzerDiagnosticsAsync(solution, options.SolutionPath, analyzers, options, telemetryCollector, cancellationToken).ConfigureAwait(true);
+                var allDiagnostics = diagnostics.Where(i=> i.Value != null && i.Value.Any()).SelectMany(i => i.Value).ToImmutableArray();
+
+                Console.WriteLine($"Found {allDiagnostics.Length} diagnostics in {stopwatch.ElapsedMilliseconds}ms");
+                telemetryCollector.WriteTelemetry();
+
+                foreach (var group in allDiagnostics.GroupBy(i => i.Id).OrderBy(i => i.Key, StringComparer.OrdinalIgnoreCase))
+                {
+                    Console.WriteLine($"  {group.Key}: {group.Count()} instances");
+
+                    // Print out analyzer diagnostics like AD0001 for analyzer exceptions
+                    if (group.Key.StartsWith("AD", StringComparison.Ordinal))
+                    {
+                        foreach (var item in group)
+                        {
+                            Console.WriteLine(item);
+                        }
+                    }
+                }
+
+                if (!string.IsNullOrWhiteSpace(options.LogFileName))
+                {
+                    WriteDiagnosticResults(diagnostics.SelectMany(i => i.Value.Select(j => Tuple.Create(i.Key, j))).ToImmutableArray(), options.LogFileName);
+                }
+
+                if (options.CodeFixes)
+                {
+                    await TestCodeFixesAsync(stopwatch, solution, analyzerAssemblies, allDiagnostics, cancellationToken).ConfigureAwait(true);
+                }
+
+                if (options.FixAll)
+                {
+                    await TestFixAllAsync(stopwatch, solution, analyzerAssemblies, diagnostics, options.ApplyChanges, cancellationToken).ConfigureAwait(true);
+                }
+            }
+        }
+        
+        private static void WriteDiagnosticResults(ImmutableArray<Tuple<ProjectId, Diagnostic>> diagnostics, string fileName)
+        {
+            var orderedDiagnostics =
+                diagnostics
+                .OrderBy(i => i.Item2.Id)
+                .ThenBy(i => i.Item2.Location.SourceTree?.FilePath, StringComparer.OrdinalIgnoreCase)
+                .ThenBy(i => i.Item2.Location.SourceSpan.Start)
+                .ThenBy(i => i.Item2.Location.SourceSpan.End);
+
+            var uniqueLines = new HashSet<string>();
+            StringBuilder completeOutput = new StringBuilder();
+            StringBuilder uniqueOutput = new StringBuilder();
+            foreach (var diagnostic in orderedDiagnostics)
+            {
+                string message = diagnostic.Item2.ToString();
+                string uniqueMessage = $"{diagnostic.Item1}: {diagnostic.Item2}";
+                completeOutput.AppendLine(message);
+                if (uniqueLines.Add(uniqueMessage))
+                {
+                    uniqueOutput.AppendLine(message);
```

Comment:
why unique is needed? worrying about 2 analyzers reporting same issue? I think this doesn't needed for our purpose?

Reply:
:arrow_right: It's a workaround for Roslyn bugs. See the **Outputs** and **Usage notes** sections of DotNetAnalyzers/StyleCopAnalyzers#1970. These cases are all expected to be fixed, but we've seen deviations from this and it may be nice to keep this feature as a preventative measure.